test/vfd_swmr_zoo_writer.c

```diff
@@ -346,44 +381,108 @@ main(int argc, char **argv)
 
         ostate = initstate(seed, vector, _arraycount(vector));
 
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
+            errx(EXIT_FAILURE, "%s: clock_gettime", __func__);
+
         if (!create_zoo(fid, ".", &lastmsgtime, config))
             errx(EXIT_FAILURE, "create_zoo didn't pass self-check");
 
-        /* Avoid deadlock: flush the file before waiting for the reader's
-         * message.
-         */
-        if (H5Fflush(fid, H5F_SCOPE_GLOBAL) < 0)
-            errx(EXIT_FAILURE, "%s: H5Fflush failed", __func__);
+        /* Notify the reader of finishing zoo creation */
+        notify = 2;
+        if (HDwrite(fd_writer_to_reader, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "write failed");
 
-        if (read(STDIN_FILENO, &step, sizeof(step)) == -1)
-            err(EXIT_FAILURE, "read");
+        /* During the wait, writer makes repeated HDF5 API calls so as to trigger
+         * EOT at approximately the correct time */
+        for(i = 0; i < swmr_config.max_lag + 1; i++) {
+            decisleep(swmr_config.tick_len);
+
+            H5Aexists(fid, "nonexistent");
+        }
 
-        if (step != 'b')
-            errx(EXIT_FAILURE, "expected 'b' read '%c'", step);
+        /* Wait until the reader finishes validating zoo creation */
+        if (HDread(fd_reader_to_writer, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "read failed");
+        if (notify != 3)
+            errx(EXIT_FAILURE, "expected 2 but read %d", notify);
+
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) == -1)
+            errx(EXIT_FAILURE, "%s: clock_gettime", __func__);
 
         if (!delete_zoo(fid, ".", &lastmsgtime, config))
             errx(EXIT_FAILURE, "delete_zoo failed");
         (void)setstate(ostate);
+
+        /* Notify the reader about finishing zoo deletion */
+        notify = 4;
+        if (HDwrite(fd_writer_to_reader, &notify, sizeof(int)) == -1)
+            err(EXIT_FAILURE, "write failed");
+
     } else {
         dbgf(2, "Reading zoo...\n");
 
+        /* Start to validate zoo creation after receiving the writer's notice */
+        if (HDread(fd_writer_to_reader, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "read failed");
+        if (notify != 1)
+            errx(EXIT_FAILURE, "unexpected message %d", notify);
+
+        /* Get the current time */
+        if (clock_gettime(CLOCK_MONOTONIC, &lastmsgtime) < 0)
+            errx(EXIT_FAILURE, "%s: clock_gettime", __func__);
+
+        last.tv_sec = lastmsgtime.tv_sec;
+        last.tv_nsec = lastmsgtime.tv_nsec;
+
         while (!validate_zoo(fid, ".", &lastmsgtime, config))
             ;
 
-        if (write(STDOUT_FILENO, &step, sizeof(step)) == -1)
-            err(EXIT_FAILURE, "write");
+        /* Make sure zoo validation doesn't take longer than the expected time */
+        if (below_speed_limit(&last, &ival))
+            warnx("validate_zoo took too long to finish");
```

Comment:
This isn't really how `below_speed_limit` is meant to be used. :-) You need a function like [`timespeccmp`](https://man.netbsd.org/timespeccmp).

Reply:
I think a new function will be similar to `below_speed_limit `